packages/git/src/common/model.ts

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+export interface WorkingDirectoryStatus {
+
+    /**
+     * `true` if the repository exists, otherwise `false`.
+     */
+    readonly exists: boolean;
+
+    /**
+     * An array of changed files.
+     */
+    readonly changes: FileChange[];
+
+    /**
+     * The optional name of the branch. Can be absent.
+     */
+    readonly branch?: string;
+
+    /**
+     * The name of the upstream branch. Optional.
+     */
+    readonly upstreamBranch?: string;
+
+    /**
+     * Wraps the `ahead` and `behind` numbers.
+     */
+    readonly aheadBehind?: { ahead: number, behind: number };
+
+    /**
+     * The hash string of the current HEAD.
+     */
+    readonly currentHead?: string;
+}
+
+export namespace WorkingDirectoryStatus {
+
+    /**
+     * `true` if the directory statuses are deep equal, otherwise `false`.
+     */
+    export function equals(left: WorkingDirectoryStatus | undefined, right: WorkingDirectoryStatus | undefined): boolean {
+        if (left && right) {
+            return left.exists === right.exists
+                && left.branch === right.branch
+                && left.upstreamBranch === right.upstreamBranch
+                && left.currentHead === right.currentHead
+                && (left.aheadBehind ? left.aheadBehind.ahead : -1) === (right.aheadBehind ? right.aheadBehind.ahead : -1)
+                && (left.aheadBehind ? left.aheadBehind.behind : -1) === (right.aheadBehind ? right.aheadBehind.behind : -1)
+                && left.changes.length === right.changes.length
+                && left.changes.sort(FileChange.compare).join(' ') === right.changes.sort(FileChange.compare).join(' ');
+        } else {
+            return left === right;
+        }
+    }
+
+    /**
+     * `true` if the status has no file changes and neither behind nor ahead from the remote branch.
+     */
+    export function isEmpty(status: WorkingDirectoryStatus): boolean {
+        return status.changes.length === 0 && (!status.aheadBehind || (status.aheadBehind.ahead === 0 && status.aheadBehind.behind === 0));
+    }
+
+}
+
+/**
+ * Enumeration of states that a file resource can have in the working directory.
+ */
+export enum FileStatus {
+    'New',
+    'Modified',
+    'Deleted',
+    'Renamed',
+    'Conflicted',
+    'Copied'
+}
+
+/**
+ * Representation of an individual file change in the working directory.
+ */
+export interface FileChange {
```

Comment:
It would be nice to prefix such common names with `Git`, there is already `FileChange`

Reply:
Or we put everything into the `Git` namespace?